pkg/cloudprovider/provider/baremetal/plugins/tinkerbell/driver.go

```diff
@@ -18,205 +18,193 @@ package tinkerbell
-	if err := json.Unmarshal(hwSpec.Raw, &hw); err != nil {
-		return nil, fmt.Errorf("failed to unmarshal tinkerbell hardware spec: %w", err)
-	}
-
-	fetchedHW, err := d.hardwareClient.Get(ctx, string(uid), hw.GetIPAddress(),
-		hw.GetMACAddress())
+func (d *driver) GetServer(ctx context.Context) (plugins.Server, error) {
+	targetHardware, err := d.HardwareClient.GetHardware(ctx, d.HardwareRef)
 	if err != nil {
-		if resourceNotFoundErr(err) {
-			return nil, cloudprovidererrors.ErrInstanceNotFound
-		}
+		return nil, err
+	}
 
-		return nil, fmt.Errorf("failed to get hardware: %w", err)
+	if targetHardware.Spec.Metadata != nil && targetHardware.Spec.Metadata.State != tinktypes.Staged {
+		return nil, cloudprovidererrors.ErrInstanceNotFound
 	}
 
-	return &HardwareSpec{
-		Hardware: tinkpkg.HardwareWrapper{
-			Hardware: fetchedHW,
-		},
-	}, nil
+	server := tinktypes.Hardware{Hardware: targetHardware}
+	return &server, nil
 }
 
-func (d *driver) ProvisionServer(ctx context.Context, uid types.UID, cfg *plugins.CloudConfigSettings, hwSpec runtime.RawExtension) (plugins.Server, error) {
-	hw := HardwareSpec{}
-	if err := json.Unmarshal(hwSpec.Raw, &hw); err != nil {
-		return nil, fmt.Errorf("failed to unmarshal tinkerbell hardware spec: %w", err)
-	}
-	hw.Hardware.Id = string(uid)
-	_, err := d.hardwareClient.Get(ctx, hw.Hardware.Id, "", "")
+func (d *driver) ProvisionServer(ctx context.Context, log *zap.SugaredLogger, meta metav1.ObjectMeta, _ runtime.RawExtension, userdata string) (plugins.Server, error) {
+	// Get the hardware object from tinkerbell
+	hardware, err := d.HardwareClient.GetHardware(ctx, d.HardwareRef)
 	if err != nil {
-		if resourceNotFoundErr(err) {
-			cfg, err := d.metadataClient.GetMachineMetadata()
-			if err != nil {
-				return nil, fmt.Errorf("failed to get metadata configs: %w", err)
-			}
-
-			hw.Hardware.Network.Interfaces[0].Dhcp.Mac = cfg.MACAddress
-
-			ip, netmask, _, err := util.CIDRToIPAndNetMask(cfg.CIDR)
-			if err != nil {
-				return nil, fmt.Errorf("failed to parse CIDR: %w", err)
-			}
-			dhcpIP := &hardware.Hardware_DHCP_IP{
-				Address: ip,
-				Netmask: netmask,
-				Gateway: cfg.Gateway,
-			}
-			hw.Hardware.Network.Interfaces[0].Dhcp.Ip = dhcpIP
-
-			if err := d.hardwareClient.Create(ctx, hw.Hardware.Hardware); err != nil {
-				return nil, fmt.Errorf("failed to register hardware to tink-server: %w", err)
-			}
+		return nil, err
+	}
+
+	var allowProvision bool
+	for _, iface := range hardware.Spec.Interfaces {
+		if iface.Netboot != nil && iface.Netboot.AllowPXE != nil && iface.Netboot.AllowPXE == ptr.Bool(false) {
+			continue
+		}
+
+		if iface.Netboot != nil && iface.Netboot.AllowWorkflow != nil && iface.Netboot.AllowWorkflow == ptr.Bool(false) {
+			continue
 		}
+
+		allowProvision = true
+	}
+
+	if !allowProvision {
+		log.Infow("server is not allowed to be provisioned; either hardware allowPXE and allowWorkflow is set tp false", "hardware", hardware.Name)
+		return nil, nil
```

Comment:
Returning nil, nil here will make the baremetal provider return a baremetalServer struct with nil in it, which will then later explode because its GetName() functions are not capable of handling nils.

If a server cannot be provisioned, it should IMHO be an error that can bubble up and not hidden in the MC log where few people would ever look (especially not the user who doesn't have access to MC logs in KKP systems).

Reply:
Good catch! I agree this makes sense